File: python/30. Substring with Concatenation of All Words.py
import copy
import collections
class Solution(object):
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        counter = collections.Counter(words)
        res= []
        wordlen = len(words[0])
        wordslen = len(words)
        slen = len(s)
        
        for i in range(wordlen):
            temp = copy.deepcopy(counter)
            count = 0
            head = i
            for j in range(head, slen - wordlen + 1, wordlen):
                tempword = s[j:j+wordlen]
                temp[tempword] -= 1
                while temp[tempword] < 0:
                    temp[s[head:head+wordlen]] += 1
                    head += wordlen
                if wordslen * wordlen == j + wordlen - head:
                    res.append(head)
        return res

File: cpp/1-50/30. Substring with Concatenation of All Words.cpp
class Solution {
public:
	vector<int> findSubstring(string s, vector<string>& words) {
		vector<int> res;
		map<string, bool> dict;
		if (words.empty())
			return res;
		int word_size = words[0].size();
		int length = words.size();
		if (s.size() < word_size * words.size())
			return res;
		for (int i = 0; i < words.size(); i++)
		{
			dict[words[i]]++;
		}
		for (int i = 0; i < s.size() - word_size * length + 1; i++)
		{
			map<string, int> cur;
			int j = 0;
			for (;j < length; j++) {
				string temp = s.substr(i + j * word_size, word_size);
				if (dict[temp]) {
					cur[temp]++;
					if (cur[temp] > dict[temp])
						break;
				}
				else break;
			}
			if(j == length)
				res.push_back(i);
		}
		return res;
	}
};
